### DISCUSSION TITLE
pnpm workspaces + tsconfig "paths" triggers "The inferred type of 'appRouter' cannot be named..."

### DISCUSSION THREAD

**Discussion by stevecaldwell77** [trpc/trpc#4301] [Q&A]
### Provide environment information

```
  System:
    OS: macOS 13.3.1
    CPU: (10) x64 Apple M1 Pro
    Memory: 44.26 MB / 32.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 14.19.3 - ~/Library/Caches/fnm_multishells/46933_1682868189959/bin/node
    Yarn: 1.22.19 - /opt/homebrew/bin/yarn
    npm: 6.14.17 - ~/Library/Caches/fnm_multishells/46933_1682868189959/bin/npm
  Browsers:
    Chrome: 112.0.5615.137
    Safari: 16.4
```

### Describe the bug

I have a monorepo using pnpm workspaces [1] where I had 2 packages setting up a TRPC server, with one package importing the other package's client. It was triggering this error when trying to build:

```
The inferred type of 'appRouter' cannot be named without a reference to '~/../package1/node_modules/@trpc/server/dist'. This is likely not portable. A type annotation is necessary.
```

I setup a blank project to try and re-create the bug with the minimal amount of code. The results are here:

  https://github.com/stevecaldwell77/trpc-bug-report

This repo has 2 packages, creatively named:

  * `packages/package1`
  * `packages/package2`

`package2` depends on `package1` and imports the client from `package1` in its router - see `packages/package2/src/app-router.ts`

When you try to build `package2`, you get the error.

From fiddling around, I narrowed the issue down to this section in `packages/package2/tsconfig.json`:

```
    "paths": {
      "~/*": ["./*"]
    },
```

If I remove that configuration, and then re-factor all the imports in `package2` to not use path aliases, it compiles fine.

But I like my path aliases!! :)

[1]: my actual project that initially triggered this bug uses [Rush](https://rushjs.io/), but I was able to re-create just using vanilla pnpm workspaces.

### Link to reproduction

https://stackblitz.com/github/stevecaldwell77/trpc-bug-report

### To reproduce

```bash
git clone https://github.com/stevecaldwell77/trpc-bug-report.git
cd trpc-bug-report  
pnpm install
(cd packages/package1 && pnpm build)
(cd packages/package2 && pnpm build)
```

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

**Comment by Nick-Lucas** [comment]
Is @trpc/server installed in both projects? This comes up a lot and it's because the client project hasn't got access to the node_module it's telling you about

**Comment by stevecaldwell77** [comment]
> Is @trpc/server installed in both projects? This comes up a lot and it's because the client project hasn't got access to the node_module it's telling you about

Yes:

 * https://github.com/stevecaldwell77/trpc-bug-report/blob/main/packages/package1/package.json
 * https://github.com/stevecaldwell77/trpc-bug-report/blob/main/packages/package2/package.json

**Comment by KATT** [comment]
This doesn't look like a bug in trpc but how it's used. Hundreds (or thousands?) of people use trpc in monorepos.

Might be that `@trpc/*` needs to be hoisted or something. Converting to discussion.
